rm segfault on network disconnect

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x50837]

goroutine 21 [running]:
main.rmList.func1(0x0, 0x0, 0x0, 0x0)
	/Users/harsha/mygo/src/github.com/minio/mc/rm-main.go:84 +0x2c7
main.rmList.func2(0xc820086138, 0xc820096840)
	/Users/harsha/mygo/src/github.com/minio/mc/rm-main.go:102 +0x2b
created by main.rmList
	/Users/harsha/mygo/src/github.com/minio/mc/rm-main.go:104 +0x22f

goroutine 1 [runnable]:
sync.(*Mutex).Unlock(0xc8203bf0bc)
	/usr/local/Cellar/go/1.5.1/libexec/src/sync/mutex.go:99
regexp.(*Regexp).put(0xc8203bf040, 0xc8203ac7e0)
	/usr/local/Cellar/go/1.5.1/libexec/src/regexp/regexp.go:212 +0x90
regexp.(*Regexp).doExecute(0xc8203bf040, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc8203eddd3, 0x8, 0x0, 0x0, ...)
	/usr/local/Cellar/go/1.5.1/libexec/src/regexp/exec.go:461 +0x206
regexp.(*Regexp).MatchString(0xc8203bf040, 0xc8203eddd3, 0x8, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/regexp/regexp.go:400 +0x67
regexp.MatchString(0x5d2e00, 0x23, 0xc8203eddd3, 0x8, 0x63ab00, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/regexp/regexp.go:427 +0x83
github.com/minio/mc/vendor/github.com/minio/minio-go-legacy.invalidBucketError.func1(0xc8203eddd3, 0x8, 0xc8203eddd3)
	/Users/harsha/mygo/src/github.com/minio/mc/vendor/github.com/minio/minio-go-legacy/errors.go:121 +0xd1
github.com/minio/mc/vendor/github.com/minio/minio-go-legacy.invalidBucketError(0xc8203eddd3, 0x8, 0x0, 0x0)
	/Users/harsha/mygo/src/github.com/minio/mc/vendor/github.com/minio/minio-go-legacy/errors.go:125 +0x85
github.com/minio/mc/vendor/github.com/minio/minio-go-legacy.apiCore.deleteObject(0xc8200a0800, 0xc8203eddd3, 0x8, 0xc8203edddc, 0x47, 0x0, 0x0)
	/Users/harsha/mygo/src/github.com/minio/mc/vendor/github.com/minio/minio-go-legacy/api-core.go:700 +0x52
github.com/minio/mc/vendor/github.com/minio/minio-go-legacy.api.RemoveObject(0xc8200a0800, 0xc8203eddd3, 0x8, 0xc8203edddc, 0x47, 0x0, 0x0)
	/Users/harsha/mygo/src/github.com/minio/mc/vendor/github.com/minio/minio-go-legacy/api.go:725 +0xf6
github.com/minio/mc/pkg/client/s3v2.(*s3Client).Remove(0xc8204bb3a0, 0x0, 0xf04c28)
	/Users/harsha/mygo/src/github.com/minio/mc/pkg/client/s3v2/s3v2.go:90 +0x1e0
main.rmAll(0x7fff5fbffae6, 0x22)
	/Users/harsha/mygo/src/github.com/minio/mc/rm-main.go:133 +0x406
```

--- a/rm-main.go
+++ b/rm-main.go
@@ -67,101 +67,121 @@ EXAMPLES:
 `,
 }
 
-func rmList(url string) (<-chan string, *probe.Error) {
+type rmListOnChannel struct {
+	keyName string
+	err     *probe.Error
+}
+
+func rmList(url string) <-chan rmListOnChannel {
+	rmListCh := make(chan rmListOnChannel)
 	clnt, err := url2Client(url)
 	if err != nil {
-		errorIf(err.Trace(), "Unable to get client object for "+url)
-		return nil, err.Trace()
+		rmListCh <- rmListOnChannel{
+			keyName: "",
+			err:     err.Trace(url),
+		}
+		return rmListCh
 	}
 	in := clnt.List(true, false)
-	out := make(chan string)
-
 	var depthFirst func(currentDir string) (*client.Content, bool)
-
 	depthFirst = func(currentDir string) (*client.Content, bool) {
 		entry, ok := <-in
 		for {
+			if entry.Err != nil {
+				rmListCh <- rmListOnChannel{
+					keyName: "",
+					err:     entry.Err,
+				}
+				return nil, false
+			}
 			if !ok || !strings.HasPrefix(entry.Content.Name, currentDir) {
 				return entry.Content, ok
 			}
 			if entry.Content.Type.IsRegular() {
-				out <- entry.Content.Name
+				rmListCh <- rmListOnChannel{
+					keyName: entry.Content.Name,
+					err:     nil,
+				}
 			}
 			if entry.Content.Type.IsDir() {
 				var content *client.Content
 				content, ok = depthFirst(entry.Content.Name)
-				out <- entry.Content.Name
-				entry = client.ContentOnChannel{Content: content}
+				rmListCh <- rmListOnChannel{
+					keyName: entry.Content.Name,
+					err:     nil,
+				}
+				entry = client.ContentOnChannel{
+					Content: content,
+					Err:     nil,
+				}
 				continue
 			}
 			entry, ok = <-in
 		}
 	}
-
 	go func() {
 		depthFirst("")
-		close(out)
+		close(rmListCh)
 	}()
-	return out, nil
+	return rmListCh
 }
 
 func rmSingle(url string) {
 	clnt, err := url2Client(url)
 	if err != nil {
-		errorIf(err.Trace(), "Unable to get client object for "+url+".")
+		errorIf(err.Trace(url), "Unable to get client object for "+url+".")
 		return
 	}
 	err = clnt.Remove(false)
-	errorIf(err.Trace(), "Unable to remove "+url+".")
+	errorIf(err.Trace(url), "Unable to remove "+url+".")
 }
 
 func rmAll(url string) {
-	out, err := rmList(url)
-	if err != nil {
-		errorIf(err.Trace(), "Unable to List "+url+".")
-		return
-	}
 	urlDir := url2Dir(url)
-	for urlPartial := range out {
+	for rmListCh := range rmList(url) {
+		if rmListCh.err != nil {
+			// if rmList throws an error die here.
+			fatalIf(rmListCh.err.Trace(), "Unable to list : "+url+" .")
+		}
 		newURL := client.NewURL(urlDir)
-		newURL.Path = filepath.Join(newURL.Path, urlPartial)
-		newclnt, e := url2Client(newURL.String())
-		if e != nil {
-			errorIf(e, "Unable to create client object : "+newURL.String()+".")
+		newURL.Path = filepath.Join(newURL.Path, rmListCh.keyName)
+		newClnt, err := url2Client(newURL.String())
+		if err != nil {
+			errorIf(err.Trace(newURL.String()), "Unable to create client object : "+newURL.String()+" .")
 			continue
 		}
-		err = newclnt.Remove(false)
-		errorIf(err, "Unable to remove : "+newURL.String()+".")
+		err = newClnt.Remove(false)
+		errorIf(err.Trace(newURL.String()), "Unable to remove : "+newURL.String()+" .")
 	}
 }
 
 func rmIncompleteUpload(url string) {
 	clnt, err := url2Client(url)
 	if err != nil {
-		errorIf(err.Trace(), "Unable to get client object for "+url+".")
+		errorIf(err.Trace(), "Unable to get client object for "+url+" .")
 		return
 	}
 	err = clnt.Remove(true)
-	errorIf(err.Trace(), "Unable to remove "+url+".")
+	errorIf(err.Trace(), "Unable to remove "+url+" .")
 }
 
 func rmAllIncompleteUploads(url string) {
 	clnt, err := url2Client(url)
 	if err != nil {
-		errorIf(err.Trace(), "Unable to get client object for "+url+".")
+		errorIf(err.Trace(url), "Unable to get client object for "+url+" .")
 		return
 	}
-	urlPartial1 := url2Dir(url)
-	ch := clnt.List(true, true)
-	for entry := range ch {
-		urlFull := filepath.Join(urlPartial1, entry.Content.Name)
-		newclnt, e := url2Client(urlFull)
-		if e != nil {
-			errorIf(e, "Unable to create client object : "+urlFull+".")
+	urlDir := url2Dir(url)
+	for entry := range clnt.List(true, true) {
+		newURL := client.NewURL(urlDir)
+		newURL.Path = filepath.Join(newURL.Path, entry.Content.Name)
+		newClnt, err := url2Client(newURL.String())
+		if err != nil {
+			errorIf(err.Trace(newURL.String()), "Unable to create client object : "+newURL.String()+" .")
 			continue
 		}
-		err = newclnt.Remove(true)
-		errorIf(err, "Unable to remove : "+urlFull+".")
+		err = newClnt.Remove(true)
+		errorIf(err.Trace(newURL.String()), "Unable to remove : "+newURL.String()+" .")
 	}
 }
 

